lxd/project/permissions.go

```diff
@@ -1419,30 +1418,36 @@ var aggregateLimitConfigValuePrinters = map[string]func(int64) string{
 
 // FilterUsedBy filters a UsedBy list based on project access.
 func FilterUsedBy(authorizer auth.Authorizer, r *http.Request, entries []string) []string {
-	// Shortcut for admins and non-RBAC environments.
-	if authorizer.UserIsAdmin(r) {
-		return entries
-	}
-
 	// Filter the entries.
 	usedBy := []string{}
 	for _, entry := range entries {
-		projectName := api.ProjectDefaultName
-
-		// Try to parse the query part of the URL.
-		u, err := url.Parse(entry)
+		entityType, projectName, pathArgs, err := cluster.URLToEntityType(entry)
 		if err != nil {
-			// Skip URLs we can't parse.
 			continue
 		}
 
-		// Check if project= is specified in the URL.
-		val := u.Query().Get("project")
-		if val != "" {
-			projectName = val
+		var object auth.Object
+		switch entityType {
+		case cluster.TypeImage:
+			object = auth.ObjectImage(projectName, pathArgs[0])
+		case cluster.TypeInstance:
+			object = auth.ObjectInstance(projectName, pathArgs[0])
+		case cluster.TypeNetwork:
+			object = auth.ObjectNetwork(projectName, pathArgs[0])
+		case cluster.TypeProfile:
+			object = auth.ObjectProfile(projectName, pathArgs[0])
+		case cluster.TypeStoragePool:
+			object = auth.ObjectStoragePool(pathArgs[0])
+		case cluster.TypeStorageVolume:
+			object = auth.ObjectStorageVolume(projectName, pathArgs[0], pathArgs[1], pathArgs[2])
+		case cluster.TypeStorageBucket:
+			object = auth.ObjectStorageBucket(projectName, pathArgs[0], pathArgs[1])
+		default:
+			continue
 		}
 
-		if !authorizer.UserHasPermission(r, projectName, "view") {
+		err = authorizer.CheckPermission(r.Context(), r, object, auth.EntitlementCanView)
```

Comment:
Can we avoid calling this function for each entity in `entries` - this could be a lot of them and slow?
Can we use the transactional checker concept we discussed?